File: src/calendar/buildFestivalsTable.ts
// Build a festival table for an arbitrary location. This is the engine-side
// counterpart to the bundled-JSON lookup in `festivalsTable.ts`.
//
// Use it to pre-compute a location-specific table once and cache it — the
// intended pattern for an offline app serving users outside IST, where
// festival dates can shift by ±1 day. Compute the table for the user's
// actual location on first use (a few seconds on-device), persist the
// returned JSON, and read it back through `getFestivalsForYear` /
// `getFestivalsForDate` by passing it as their `source` argument.

import { computeFestivalsInRange } from './yearly';
import type { GeoLocation } from '../types/location';
import type { AyanamsaType, MasaSystem, FestivalRegion } from '../types/options';
import type {
  FestivalDictEntry,
  FestivalsFile,
  FestivalsTableLanguage,
  LocalizedString,
  PackedFestivalTableDay,
} from './festivalsTableTypes';

export interface BuildFestivalsTableOptions {
  /** Observer coordinates. */
  location: GeoLocation;
  /**
   * UTC offset in minutes east of UTC, e.g. `330` for IST, `-300` for US
   * Eastern (EST), `0` for UK/GMT. Used both to compute the panchang and to
   * group emissions into local calendar dates; stamped into `_meta`.
   */
  timezoneOffsetMinutes: number;
  /** First Gregorian year to include (inclusive). */
  startYear: number;
  /** Last Gregorian year to include (inclusive). */
  endYear: number;
  /** Locales to emit. Defaults to `['en', 'hi']`. */
  languages?: readonly FestivalsTableLanguage[];
  /** Sidereal system. Defaults to `'lahiri'`. */
  ayanamsa?: AyanamsaType;
  /** Lunar month naming. Defaults to `'purnimanta'`. */
  masaSystem?: MasaSystem;
  /** Regional scope. Defaults to `'all'`. */
  region?: FestivalRegion;
  /** Human-readable label for the location, stamped into `_meta`. */
  referenceLocation?: string;
  /** ISO timestamp stamped into `_meta.generatedAt`. Defaults to `''`. */
  generatedAt?: string;
  /** Free-text note stamped into `_meta.note`. */
  note?: string;
}

const DEFAULT_NOTE =
  'Pre-computed festival table. Eclipses are excluded because visibility ' +
  'is location-dependent; use getUpcomingEclipses for those.';

function toDateKey(d: Date, offsetMinutes: number): string {
  const shifted = new Date(d.getTime() + offsetMinutes * 60_000);
  const y = shifted.getUTCFullYear();
  const m = String(shifted.getUTCMonth() + 1).padStart(2, '0');
  const day = String(shifted.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

/**
 * Interns unique festival descriptors so each day can hold indices instead of
 * repeating every localized string at every occurrence.
 *
 * The whole `(key, type, name, description)` tuple is the dedup unit, not the
 * name alone: a festival's description is not a pure function of its key —
 * Raksha Bandhan carries a different note depending on whether Bhadra overlaps
 * the day — so keying on the name would merge entries that differ.
 */
class FestivalDictionary {
  readonly entries: FestivalDictEntry[] = [];
  private readonly index = new Map<string, number>();

  intern(entry: FestivalDictEntry): number {
    const id = JSON.stringify([entry.key, entry.type, entry.name, entry.description ?? null]);
    const seen = this.index.get(id);
    if (seen !== undefined) return seen;
    const next = this.entries.length;
    this.entries.push(entry);
    this.index.set(id, next);
    return next;
  }
}

/**
 * Compute a festival table for the given location and year range.
 *
 * Runs the festival engine once per requested language and zips the runs
 * by emission index (festival selection is language-independent, so the
 * runs align). Eclipses are dropped — their visibility is location-specific
 * and better served by `getUpcomingEclipses`.
 *
 * The emitted table is dictionary-encoded and carries each festival's stable
 * `key`; see `festivalsTableTypes.ts` for why.
 *
 * @returns A {@link FestivalsFile} ready to serialize, cache, and feed back
 *          into the `readFestivalsForYear` / `readFestivalsForDate` accessors
 *          via their `source` argument.
 */
export function buildFestivalsTable(
  opts: BuildFestivalsTableOptions,
): FestivalsFile {
  const {
    location,
    timezoneOffsetMinutes,
    startYear,
    endYear,
    languages = ['en', 'hi'],
    ayanamsa = 'lahiri',
    masaSystem = 'purnimanta',
    region = 'all',
    referenceLocation = '',
    generatedAt = '',
    note = DEFAULT_NOTE,
  } = opts;

  if (!Number.isInteger(startYear) || !Number.isInteger(endYear)) {
    throw new RangeError('startYear and endYear must be integers');
  }
  if (startYear > endYear) {
    throw new RangeError(`startYear (${startYear}) must be ≤ endYear (${endYear})`);
  }
  if (languages.length === 0) {
    throw new RangeError('languages must contain at least one locale');
  }

  const years: Record<string, PackedFestivalTableDay[]> = {};
  // One dictionary across the whole table, not one per year: most festivals
  // recur annually, so sharing it is where the compression comes from.
  const dict = new FestivalDictionary();

  for (let year = startYear; year <= endYear; year++) {
    years[String(year)] = buildYear(
      year, location, timezoneOffsetMinutes, languages,
      ayanamsa, masaSystem, region, dict,
    );
  }

  return {
    _meta: {
      format: 2,
      referenceLocation,
      latitude: location.latitude,
      longitude: location.longitude,
      timezoneOffsetMinutes,
      ayanamsa,
      masaSystem,
      region,
      languages: [...languages],
      startYear,
      endYear,
      generatedAt,
      note,
    },
    _dict: dict.entries,
    years,
  };
}

function buildYear(
  year: number,
  location: GeoLocation,
  offsetMinutes: number,
  languages: readonly FestivalsTableLanguage[],
  ayanamsa: AyanamsaType,
  masaSystem: MasaSystem,
  region: FestivalRegion,
  dict: FestivalDictionary,
): PackedFestivalTableDay[] {
  const start = new Date(Date.UTC(year, 0, 1));
  const end = new Date(Date.UTC(year, 11, 31));

  // One engine run per locale; selection is locale-independent so the runs
  // emit the same festivals in the same order — we zip them by index.
  const runs = languages.map(language =>
    computeFestivalsInRange(start, end, location, {
      timezone: offsetMinutes,
      ayanamsa,
      masaSystem,
      region,
      language,
    }),
  );

  const len = runs[0]!.length;
  for (let i = 1; i < runs.length; i++) {
    if (runs[i]!.length !== len) {
      throw new Error(
        `language-run length mismatch for ${year}: ` +
        `${languages[0]}=${len} ${languages[i]}=${runs[i]!.length}`,
      );
    }
  }

  const byDate = new Map<string, number[]>();
  for (let i = 0; i < len; i++) {
    const base = runs[0]![i]!;
    if (base.festival.type === 'eclipse') continue;

    const name: LocalizedString = {};
    let description: LocalizedString | undefined;
    for (let l = 0; l < languages.length; l++) {
      const f = runs[l]![i]!.festival;
      name[languages[l]!] = f.name;
      if (f.description) {
        description ??= {};
        description[languages[l]!] = f.description;
      }
    }

    const entry: FestivalDictEntry = {
      key: base.festival.key,
      type: base.festival.type,
      name,
    };
    if (description) entry.description = description;

    const dateKey = toDateKey(base.date, offsetMinutes);
    let bucket = byDate.get(dateKey);
    if (!bucket) {
      bucket = [];
      byDate.set(dateKey, bucket);
    }
    bucket.push(dict.intern(entry));
  }

  return [...byDate.keys()]
    .sort()
    .map(date => ({ date, festivals: byDate.get(date)! }));
}

File: source/go/internal/calendar/buildfestivalstable.go
package calendar

import (
	"encoding/json"
	"sort"
	"strconv"
	"strings"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/core"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/types"
)

type BuildFestivalsTableOptions struct {
	Location              types.GeoLocation
	TimezoneOffsetMinutes int
	StartYear             int
	EndYear               int
	Languages             []FestivalsTableLanguage
	Ayanamsa              types.AyanamsaType
	MasaSystem            types.MasaSystem
	Region                types.FestivalRegion
	ReferenceLocation     string
	GeneratedAt           string
	Note                  *string
	RegionAliasWarner     core.RegionAliasWarner
}

const DefaultFestivalsNote = "Pre-computed festival table. Eclipses are excluded because visibility " +
	"is location-dependent; use getUpcomingEclipses for those."

type festivalDictionary struct {
	entries []FestivalDictEntry
	index   map[string]int
}

func newFestivalDictionary() *festivalDictionary {
	return &festivalDictionary{entries: []FestivalDictEntry{}, index: map[string]int{}}
}

func (d *festivalDictionary) intern(entry FestivalDictEntry) int {
	tuple := []any{entry.Key, entry.Type, entry.Name, entry.Description}
	raw, err := json.Marshal(tuple)
	if err != nil {
		panic("calendar: interning a festival entry: " + err.Error())
	}
	id := string(raw)
	if seen, ok := d.index[id]; ok {
		return seen
	}
	next := len(d.entries)
	d.entries = append(d.entries, entry)
	d.index[id] = next
	return next
}

func BuildFestivalsTable(
	ctx *astronomy.EphemerisCtx, opts BuildFestivalsTableOptions,
) (FestivalsFile, error) {
	languages := opts.Languages
	if languages == nil {
		languages = AllTableLanguages
	}
	ayanamsa := opts.Ayanamsa
	if ayanamsa == "" {
		ayanamsa = types.Lahiri
	}
	masaSystem := opts.MasaSystem
	if masaSystem == "" {
		masaSystem = types.Purnimanta
	}
	region := opts.Region
	if region == "" {
		region = types.RegionAll
	}
	note := DefaultFestivalsNote
	if opts.Note != nil {
		note = *opts.Note
	}

	if opts.StartYear > opts.EndYear {
		return FestivalsFile{}, types.Codef(types.ErrInvalidInput,
			"startYear (%d) must be ≤ endYear (%d)", opts.StartYear, opts.EndYear)
	}
	if len(languages) == 0 {
		return FestivalsFile{}, types.Codef(types.ErrInvalidInput,
			"languages must contain at least one locale")
	}

	years := map[string][]PackedFestivalTableDay{}
	dict := newFestivalDictionary()

	for year := opts.StartYear; year <= opts.EndYear; year++ {
		days, err := buildFestivalYear(ctx, year, opts.Location, opts.TimezoneOffsetMinutes,
			languages, ayanamsa, masaSystem, region, opts.RegionAliasWarner, dict)
		if err != nil {
			return FestivalsFile{}, err
		}
		years[strconv.Itoa(year)] = days
	}

	langs := make([]FestivalsTableLanguage, len(languages))
	copy(langs, languages)
	return FestivalsFile{
		Meta: FestivalTableMeta{
			Format:                2,
			ReferenceLocation:     opts.ReferenceLocation,
			Latitude:              opts.Location.Latitude,
			Longitude:             opts.Location.Longitude,
			TimezoneOffsetMinutes: opts.TimezoneOffsetMinutes,
			Ayanamsa:              string(ayanamsa),
			MasaSystem:            string(masaSystem),
			Region:                string(region),
			Languages:             langs,
			StartYear:             opts.StartYear,
			EndYear:               opts.EndYear,
			GeneratedAt:           opts.GeneratedAt,
			Note:                  note,
		},
		Dict:  dict.entries,
		Years: years,
	}, nil
}

func buildFestivalYear(
	ctx *astronomy.EphemerisCtx,
	year int,
	location types.GeoLocation,
	offsetMinutes int,
	languages []FestivalsTableLanguage,
	ayanamsa types.AyanamsaType,
	masaSystem types.MasaSystem,
	region types.FestivalRegion,
	warner core.RegionAliasWarner,
	dict *festivalDictionary,
) ([]PackedFestivalTableDay, error) {
	startMs := types.DateUTC(year, 0, 1).Ms()
	endMs := types.DateUTC(year, 11, 31).Ms()

	runs := make([][]FestivalDay, len(languages))
	for i, language := range languages {
		run, err := ComputeFestivalsInRange(ctx, startMs, endMs, location, YearlyListingOptions{
			Timezone:          types.TimezoneOffset(offsetMinutes),
			Ayanamsa:          ayanamsa,
			MasaSystem:        masaSystem,
			Region:            region,
			Language:          types.Language(language),
			RegionAliasWarner: warner,
		})
		if err != nil {
			return nil, err
		}
		runs[i] = run
	}

	length := len(runs[0])
	for i := 1; i < len(runs); i++ {
		if len(runs[i]) != length {
			return nil, types.Codef(types.ErrInvalidInput,
				"language-run length mismatch for %d: %s=%d %s=%d",
				year, languages[0], length, languages[i], len(runs[i]))
		}
	}

	byDate := map[string][]int{}
	order := []string{}
	for i := 0; i < length; i++ {
		base := runs[0][i]
		if base.Festival.Type == types.FestivalEclipse {
			continue
		}

		namePairs := make([]LocalizedPair, 0, len(languages))
		descPairs := make([]LocalizedPair, 0, len(languages))
		for l := 0; l < len(languages); l++ {
			f := runs[l][i].Festival
			namePairs = append(namePairs, LocalizedPair{Lang: languages[l], Value: f.Name})
			if f.Description != "" {
				descPairs = append(descPairs, LocalizedPair{Lang: languages[l], Value: f.Description})
			}
		}

		entry := FestivalDictEntry{
			Key:  base.Festival.Key,
			Type: FestivalsTableType(base.Festival.Type),
			Name: Localized(namePairs...),
		}
		if len(descPairs) > 0 {
			d := Localized(descPairs...)
			entry.Description = &d
		}

		dateKey := tableDateKey(base.Date.Ms(), offsetMinutes)
		if _, seen := byDate[dateKey]; !seen {
			order = append(order, dateKey)
		}
		byDate[dateKey] = append(byDate[dateKey], dict.intern(entry))
	}

	sorted := make([]string, len(order))
	copy(sorted, order)
	sort.Slice(sorted, func(i, j int) bool {
		return strings.Compare(sorted[i], sorted[j]) < 0
	})
	out := make([]PackedFestivalTableDay, 0, len(sorted))
	for _, date := range sorted {
		out = append(out, PackedFestivalTableDay{Date: date, Festivals: byDate[date]})
	}
	return out, nil
}
